src/buildstream_plugins/sources/cargo.py

```diff
@@ -148,7 +148,25 @@ def stage(self, directory):
         try:
             mirror_file = self._get_mirror_file()
             with tarfile.open(mirror_file) as tar:
-                tar.extractall(path=directory)
+
+                def is_within_directory(directory, target):
+                    abs_directory = os.path.abspath(directory)
+                    abs_target = os.path.abspath(target)
+
+                    prefix = os.path.commonprefix([abs_directory, abs_target])
+
+                    return prefix == abs_directory
+
+                def safe_extract(tar, path=".", members=None, *, numeric_owner=False):
+
+                    for member in tar.getmembers():
+                        member_path = os.path.join(path, member.name)
+                        if not is_within_directory(path, member_path):
+                            raise Exception("Attempted Path Traversal in Tar File")
+
+                    tar.extractall(path, members, numeric_owner=numeric_owner)
```

Comment:
My understanding of the `tar` is that it contains no global header, and thus doing a `tar.getmembers()` then a `tar.extracall()` actually  ends up reading the file twice, which is wasteful. If we want to sanitize I would rather extract each entry one by one after we've sanitized it's destination

Reply:
The fundamental problem here IMHO is we are handling sources in host without any kind of sandboxing or isolation. This is just duck-taping around that.

Reply:
@BenjaminSchubert `TarFile.extractall()` internally does the equivalent to `TarFile.getmembers()` anyway if no `members` argument is passed. However, it caches the member list including the offsets.

@nanonyme While I'd like to see fully sandboxed source plugins, that would be a major undertaking, clearly out of scope of this fix.